### PR TITLE
Add some basic server stats

### PR DESCRIPTION
**Pull request by Flohack74**
* Will be useful for load evaluation and sizing of the server

### CODE REVIEW

File: logger/logger.go
Comment (mardy) [inline logger/logger.go:85]: Why this change?

File: server/dev/server.go
Comment (mardy) [inline server/dev/server.go:73]: Why this change?

File: server/statistics/statistics.go
Comment (mardy) [inline server/statistics/statistics.go]: Please reformat this file according to the style of the other files in the project: use spaces instead of tabs, remove empty lines after opening braces and before closing braces, etc.

File: server/statistics/statistics.go
Comment (mardy) [inline server/statistics/statistics.go]: This is wrong: you should not reset the timers, but rather purge the older devices (and for that, you need to keep track of the time when IncreaseDevices() was called. For example, suppose that a new device gets added every 6 minutes: the stats for "60 mins" should always report 10 devices (plus or minus one), while after one hour of activity your implementation would return 0.

**Review by mardy — CHANGES_REQUESTED** [review]

File: server/dev/server.go
Comment (dpniel) [inline server/dev/server.go]: It would be cleaner to move this goroutine inside the NewStatistics constructor. It's more of an implementation detail that is better hidden inside the statistics package.

File: logger/logger.go
Comment (Flohack74) [reply logger/logger.go:85]: Because the push server logs are very wide, as the system journal adds another timestamp in front. And for a push service we do not need that precision :) - Just needs to be checked on the client side, if it is the same or we loose fidelity.

File: server/dev/server.go
Comment (Flohack74) [reply server/dev/server.go:73]: Push server logs otherwise IP addresses and tokens and the board decided that we do not want to have this information in the logs for privacy reason. For debugging purposes during the first weeks it might be necessary to turn it on again. Can we switch this dynamically during runtime with Go?

File: server/statistics/statistics.go
Comment (Flohack74) [reply server/statistics/statistics.go]: Its a different interpretation: This is not the total device count but how many devices were added or removed. Eg here:
grafik

as you see the value also can get negative. For me as a infrastructure guy its more important how much the server is being challenged with adding or removing devices, rather than the static number. But we can add this too

File: server/dev/server.go
Comment (Flohack74) [reply server/dev/server.go]: Ok yes agreed.

File: server/statistics/statistics.go
Comment (mardy) [reply server/statistics/statistics.go]: Yes, but I'm afraid it doesn't change the fact that most of the data you print is wrong: the "5 mins" line is the only one correct.
Please follow my example: if one device is added every 6 minutes, and no one is removed, what should the  "60 mins" line print, after 65 minutes of server activity? 10 or 0?

File: server/statistics/statistics.go
Comment (Flohack74) [reply server/statistics/statistics.go]: Are you sure there are only spaces used? MC reports tabs here:
grafik

Comment (mardy) [reply server/statistics/statistics.go]: Wow, you are right, this project is a mess :-(
Well, feel free to pick to coding standard you prefer then, but please make it consistent: use either tabs, or spaces (I strongly recommend using spaces :-) ).

File: server/statistics/statistics.go
Comment (Flohack74) [reply server/statistics/statistics.go]: Ok now I got you ;) - well currently its not a rolling stats so it will get cleared at minute 60 and therefore will print 0 at minute 65. At minute 66, in your point of view, it should rather print 9 right? That would roll out the first device added at minute 0 lets say...

Comment (dpniel) [reply server/statistics/statistics.go]: Could you not use an expiring counter? something like this https://paste.ubuntu.com/26219933/ (note this is not tested) You could new up an statistics counter for each period you want to track and let the janitor expire stats as you go. THen you just need to increment each counter with an expiry time and call count() when you want the running count. Just an idea...

File: server/dev/server.go
Comment (mardy) [reply server/dev/server.go:73]: I'm a bit ignorant on Go, I believe that there should be some way to control this via environment variables, but I don't know.
I think the reason you mentioned is very valid, so feel free to keep your change the way it is.

File: server/statistics/statistics.go
Comment (Flohack74) [reply server/statistics/statistics.go]: I think its much easier made with a ring typed element for all stats except the 5mins stats. Then, you just populate the ring with every tick, and sum up all elements. As soon as the ring gets full the last element will be evicted automatically.

**Comment by Flohack74** [comment]
Guys I refactored it completely, maybe its better like this now ;)

File: server/statistics/statistics.go
Comment (dpniel) [inline server/statistics/statistics.go]: **[Nitpick]** I think it would be nicer to declare these in the `&StatsValue{}` constructor. Or if you prefer them this way then move the `}` on line 20 to 19

File: server/statistics/statistics.go
Comment (dpniel) [inline server/statistics/statistics.go]: It might be nicer to make this a const variable with a descriptive name. It's a little hard to figure out from first glance what this value represents.

**Review by dpniel — APPROVED** [review]
Just a couple of nitpicks but looks good to me apart from that.

File: server/statistics/statistics.go
Comment (mardy) [reply server/statistics/statistics.go]: Actually, is this variable used at all?

File: server/statistics/statistics.go
Comment (mardy) [inline server/statistics/statistics.go]: please fix the indentation here

File: server/statistics/statistics.go
Comment (mardy) [inline server/statistics/statistics.go]: remove this line (or add a comment explaining when the comment must be removed)

File: server/statistics/statistics.go
Comment (mardy) [inline server/statistics/statistics.go]: Are you sure you need to do these initializations? I would think that they are not needed: the rings can be kept empty, and filled as you get the data.

**Review by mardy — CHANGES_REQUESTED** [review]

File: server/statistics/statistics.go
Comment (Flohack74) [reply server/statistics/statistics.go]: No its obsolete now. Removed ;)

File: server/statistics/statistics.go
Comment (Flohack74) [reply server/statistics/statistics.go]: Yes true, I was thinking that this creates a problem but it could work...

File: server/statistics/statistics.go
Comment (mardy) [inline server/statistics/statistics.go]: Do you want to keep it enabled, or is it better to have it commented out?

**Review by mardy — APPROVED** [review]
This is fine for me now. Note that I've added one comments, as I believe you accidentally uncommented one line which is for testing only.

Please also wait for Daniel's feedback before merging; and before merging, please squash all the commits into a single one (ask me in telegram, if you need help with git)

**Review by dpniel — APPROVED** [review]
LGTM

File: server/statistics/statistics.go
Comment (Flohack74) [reply server/statistics/statistics.go]: Damn didnt see this before merge...

**Pull request merged**
